csrc/id_graphs.h

```diff
@@ -0,0 +1,501 @@
+  // MappingMode. If the expressions are determined the same then
+  // if forward
+  //   will map outputs
+  // else
+  //   will map inputs
+  // in the provided mode.
+  // Returns if expressions were mapped through.
+  bool mapThroughExpr(Expr* first, Expr* second, bool forward);
+
+  // Map through loop swizzles, as input/output IterDomains are exact, only the
+  // order they're traversed differs.
+  void mapThroughLoopSwizzles();
+
+  // Maps iter domain pairs returned by calling that return mappings from
+  // IdGraph::isTrivialExpr on every expression in the graph.
+  void mapThroughTrivialExprs();
+
+  // Removes expressions from unique_definitions_ and unique_uses_ that return
+  // mappings from IdGraph::isTrivialExpr
+  void removeTrivialExprs();
+
+ private:
+  // Removes the provided expression group from unique_definitions_ and
+  // unique_uses_ breaking traversal through them.
+  void eraseExprGroup(ExprGroup expr_group);
+
+  // Keeps a disjoint set entry for all IterDomain for all mapping mode types.
+  //
+  // Using an array here might be nice, but it seems hard to use an enum as an
+  // array key
+  // https://stackoverflow.com/questions/2102582/how-can-i-count-the-items-in-an-enum
+  DisjointSets<IterDomain*> disjoint_ids_;
+
+  // Keeps a disjoint set entry for all Expressions for all mapping mode types.
+  DisjointSets<Expr*> disjoint_exprs_;
+
+  std::unordered_map<IdGroup, ExprGroups> unique_definitions_;
+
+  std::unordered_map<IdGroup, ExprGroups> unique_uses_;
+
+  // Hold a set of IterDomains that are considered view rfactor ids. This
+  // identification is particularly important to understand if split operations
+  // are divisible or not.
+  //
+  // TODO: This should just be in IterDomainGraphs, not here.
+  std::unordered_set<IterDomain*> view_rfactor_ids_;
+};
+
+// Iterates through an IterDomain Graph in topological order, calling handle on
+// all Id and all Expr groups in a forward topological order.
+//
+// Warning: Expr groups that have an input and output in the same IdGroup are
+// ignored.
+//
+// Warning: This is not a great iterator if there's a desire to minimize paths
+// traveled to simply visit all IdGroups in order. See ExprsBetween to see how
+// we might minimize paths.
+class TORCH_CUDA_CU_API IdGraphVisitor {
+ protected:
+  // If sub_selection is assumed to be a set of iter domains by which form a
+  // sub-regrion of the IdGraph provided. Only that sub-region will be visited.
+  IdGraphVisitor(
+      IdGraph& id_graph,
+      const VectorOfUniqueEntries<IterDomain*> sub_selection = {})
+      : id_graph_(id_graph), sub_selection_(sub_selection) {}
+
+  virtual void handle(IdGroup id_group) = 0;
+  virtual void handle(ExprGroup expr_group) = 0;
+
+  void traverse();
+
+  IdGraph& graph() {
+    return id_graph_;
+  };
+
+  IdGraphVisitor() = delete;
+
+  IdGraphVisitor(const IdGraphVisitor& other) = default;
+  IdGraphVisitor& operator=(const IdGraphVisitor& other) = default;
```

Comment:
I get a compilation error with this line as well as the below move assignment (I'm using clang++). 

```
/home/nmaruyama/pytorch/debug2/nvfuser/csrc/id_graphs.h:239:19: error: explicitly defaulted copy assignment operator is implicitly deleted [-Werror,-Wdefaulted-function-deleted]
  IdGraphVisitor& operator=(IdGraphVisitor& other) = default;
                  ^
/home/nmaruyama/pytorch/debug2/nvfuser/csrc/id_graphs.h:247:12: note: copy assignment operator of 'IdGraphVisitor' is implicitly deleted because field 'id_graph_' is of reference type 'nvfuser::IdGraph &'
  IdGraph& id_graph_;
           ^
/home/nmaruyama/pytorch/debug2/nvfuser/csrc/id_graphs.h:242:19: error: explicitly defaulted move assignment operator is implicitly deleted [-Werror,-Wdefaulted-function-deleted]
  IdGraphVisitor& operator=(IdGraphVisitor&& other) = default;
                  ^
/home/nmaruyama/pytorch/debug2/nvfuser/csrc/id_graphs.h:247:12: note: move assignment operator of 'IdGraphVisitor' is implicitly deleted because field 'id_graph_' is of reference type 'nvfuser::IdGraph &'
  IdGraph& id_graph_;
```

It seems this is because assigning to a reference field is not valid. https://stackoverflow.com/a/26946286

Reply:
Thanks, just mindlessly added those. I'll delete the copy assignment and move assignment. I don't really even think any of the constructors are even that useful, could probably just delete them all even.